OKR/PACKAGE-MANAGERS.md

```diff
@@ -2,6 +2,224 @@
+* Add: Package _maintainers_ won't touch IPFS if it takes several days to import the package repository.
+* Fetch: Package manager _users_ won't use IPFS to download their packages if it's slower and/or consumes significantly more bandwidth than HTTP.
+
+TODO: We could phrase this as something like "ipfs is a viable transport for package managers."
+
+#### Key Results
+
+By order of priority,
+
+* P0: Update - IPFS can quickly update a large package repository.
+* P1: Fetch - IPFS can quickly and efficiently fetch packages from a large package repository.
+* P2: Add -  IPFS can quickly add a large package repository.
+
+**Rational**
+
+Exploration in Q2 demonstrated that we could improve add performance by a factor of 2 (at least) but that factor disappeared on HDDs. While there's still some experimentation to be done, it's not clear that we can improve go-ipfs performance to a point where an _entire_ large package repository can be re-added whenever it changes.
+
+While package repositories are large, they tend not to change _all_ that much on a daily basis. We can side-step the add performance issues by ensuring that package maintainers can easily and efficiently _update_ package repositories stored in IPFS.
+
+Finally, while we've been focusing on how to ensure IPFS meets the needs of package _maintainers_, we've neglected the end user. It's unlikely that a package manager would choose to _exclusively_ use IPFS and end-users won't put up with IPFS unless it's at least as fast as HTTP and reasonably efficient. This came up when one of our partners found that using IPFS to distribute docker images within a fast datacenter network was no faster than going directly to docker-hub.
+
+**Not Included**
+
+* IPNS: To build a fully decentralized package manager, IPNS is a must. However, to get in the door with existing package managers, IPFS just needs to be a viable _transport_.
+* Performant GC: Package maintainers often keep old packages around (TODO: check this) so GC isn't all that important. Package manager _users_ can throw away their local repo when it grows too large (i.e., treat it like a cache).
+* Pinning Usability (named pins, etc.): While a pain-point and embarrassing UX issue, pinning and pin management is not a show-stopper for package managers.
+
+##### P0: Update (P0)
+
+IPFS can quickly update a large package repository.
+
+TODO: "quickly" should be defined in terms of normal package update procedures used by package maintainers for some package manager. This could be:
+
+* Updating from one snapshot to the next.
+* Mounting the repository locally and having the package maintenance tools operate on it directly.
+
+##### P1: Fetch
+
+IPFS can quickly and efficiently fetch packages from a large package repository.
+
+TODO: define "quickly and efficiently" in terms of beating HTTP.
+
+TODO: consider breaking this into two KRs? In the past, we had one for efficiency and one for throughput.
```

Comment:
1. IPFS uses at most 10% more bandwidth than HTTP.
2. IPFS maxes out a 1GiB link on the internet (100ms latencies).

Reply:
I assume we already know this, but was at the ICN conference this past week and a paper was presented (https://conferences.sigcomm.org/acm-icn/2019/proceedings/icn19-34.pdf) that shows (in go-ipfs 0.4.20) a lot of over requesting of data from Bitswap on a small closed network of 10 nodes.

Is the 10% you mention between 2 peers, or even in a multi-peer environment?

Reply:
I think steven's comment is the KR goal - not what we currently hit...